### PR TITLE
Fix issue 191

### LINKED ISSUE #191: Keep the default values of port in ChainX 2.0
In ChainX 1.0, we use these default values:

- 8086 for rpc port
- 8087 for ws port
- 20222 for libp2p port

We should also keep them in ChainX 2.0.

### PR DESCRIPTION
**Pull request by alannotnerd**
Fixes #191

### CODE REVIEW

File: cli/src/config.rs [hunk 1]
```diff
@@ -23,7 +23,7 @@ fn read_config_file(path: &Path) -> Result<HashMap<String, Value>, Box<dyn std::
 /// Only the options that do not appear in the command line will be appended.
 fn extend_cli_args(
     cli_args: Vec<String>,
-    path: &std::path::Path,
+    optional_path: Option<&std::path::Path>,
```
Comment (koushiro) [inline cli/src/config.rs]:
```suggestion
    path: Option<&Path>,
```

File: cli/src/config.rs [hunk 2]
```diff
@@ -34,42 +34,55 @@ fn extend_cli_args(
...
-            }
-            Value::Null => {}
-            Value::Object(_) => {
-                panic!("The nested configuration in the config file is unsupported.")
             }
         }
     }
 
     let mut args = cli_args;
+    for (key_, value) in port_opts {
```
Comment (koushiro) [inline cli/src/config.rs:83]:
```suggestion
    for (key, value) in port_opts {
```

File: cli/src/config.rs [hunk 3]
```diff
@@ -34,42 +34,55 @@ fn extend_cli_args(
 
     let mut config_opts = Vec::new();
 
-    for (key, value) in read_config_file(path)?.into_iter() {
-        let opt = format!("--{}", key);
+    let mut port_opts = ::std::collections::HashMap::new();
```
Comment (koushiro) [inline cli/src/config.rs:37]:
```suggestion
    let mut port_opts = HashMap::new();
```

File: cli/src/config.rs [hunk 4]
```diff
@@ -34,42 +34,55 @@ fn extend_cli_args(
...
-            Value::Object(_) => {
-                panic!("The nested configuration in the config file is unsupported.")
             }
         }
     }
 
     let mut args = cli_args;
+    for (key_, value) in port_opts {
+        config_opts.push(format!("--{}={}", key_, value));
+    }
```
Comment (koushiro) [inline cli/src/config.rs]:
```suggestion
    for (key, value) in port_opts {
        config_opts.push(format!("--{}={}", key, value));
    }
```

File: cli/src/config.rs [hunk 5]
```diff
@@ -34,42 +34,55 @@ fn extend_cli_args(
 
     let mut config_opts = Vec::new();
 
-    for (key, value) in read_config_file(path)?.into_iter() {
-        let opt = format!("--{}", key);
+    let mut port_opts = ::std::collections::HashMap::new();
+    port_opts.insert("port", 20222);
+    port_opts.insert("rpc-port", 8086);
+    port_opts.insert("ws-port", 8087);
```
Comment (koushiro) [inline cli/src/config.rs]:
```suggestion
    let mut port_opts = HashMap::new();
    port_opts.insert("port", 20222);
    port_opts.insert("rpc-port", 8086);
    port_opts.insert("ws-port", 8087);
```

Comment (koushiro) [reply cli/src/config.rs]: I recommend passing the default opts from the argument of `extend_cli_args`.

maybe like:

```
fn extend_cli_args(
    cli_args: Vec<String>,
    path: Option<&Path>,
    default_opts: HashMap<String, String>,
)  -> Result<Vec<String>, Box<dyn std::error::Error>> {...}

fn default_opts() -> HashMap<String, String> {
    let mut opts = HashMap::new();
    map.insert("port".into(), 20222.to_string());
    map.insert("rpc-port".into(), 8086.to_string());
    map.insert("ws-port".into(), 8087.to_string());
    map
}
```

File: cli/src/config.rs [hunk 6]
```diff
@@ -93,11 +106,14 @@ pub fn preprocess_cli_args(cli_args: Vec<String>) -> Vec<String> {
...
             Ok(args) => args,
             Err(e) => panic!(e.to_string()),
         }
     } else {
-        cli_args
+        match extend_cli_args(cli_args, None) {
+            Ok(args) => args,
+            Err(e) => panic!(e.to_string()),
+        }
     }
```
Comment (koushiro) [inline cli/src/config.rs]:
```
let path = config_path.as_ref().map(Path::new);
match extend_cli_args(cli_args, path) {
    Ok(args) => args,
    Err(e) => panic!(e.to_string()),
}
```

**Review by koushiro — COMMENTED** [review]
Rest LGTM

File: cli/src/config.rs (hunk 1 above)
Comment (alannotnerd) [reply cli/src/config.rs]: the destructuring `if-let` assignment below take the `path` identifier, though it's also work.  Shall I rename the variable?

File: cli/src/config.rs [hunk 7]
```diff
@@ -92,12 +110,10 @@ pub fn preprocess_cli_args(cli_args: Vec<String>) -> Vec<String> {
...
-            Err(e) => panic!(e.to_string()),
-        }
-    } else {
-        cli_args
+    let path: Option<&Path> = config_path.as_ref().map(Path::new);
+    match extend_cli_args(cli_args, path, default_opts()) {
+        Ok(args) => args,
+        Err(e) => panic!(e.to_string()),
     }
+
```
Comment (koushiro) [inline cli/src/config.rs]:
```suggestion

```

**Review by koushiro — APPROVED** [review]

File: cli/src/config.rs [hunk 8]
```diff
@@ -33,47 +34,64 @@ fn extend_cli_args(
...
-                }));
-            }
-            Value::Null => {}
-            Value::Object(_) => {
-                panic!("The nested configuration in the config file is unsupported.")
             }
         }
     }
 
     let mut args = cli_args;
```
Comment (liuchengxu) [inline cli/src/config.rs]:
```Rust
    for (key, value) in filtered_default_opts {
        config_opts.push(format!("--{}={}", key, value));
    }

    let mut args = cli_args;
```

**Pull request merged**